File: 416-partition-equal-subset-sum/partition-equal-subset-sum.py
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        summ = 0
        for elem in nums:
            summ += elem
        if summ % 2 == 1:
            return False
        else:
            dp = set()
            dp.add(0)
            for num in nums:
                dp_2 = set()
                for d in dp:
                    dp_2.add(d + num)
                    dp_2.add(d)
                dp = dp_2
        if (summ / 2) in dp:
            return True
        else:
            return False
   

File: 416-partition-equal-subset-sum/partition-equal-subset-sum.java
class Solution {
    public boolean canPartition(int[] nums) {
        int summ = 0;
        for (int num : nums){
            summ += num;
        }
        if (summ % 2 != 0){
            return false;
        }
        Set<Integer> dp = new HashSet<>();
        dp.add(0);
        for (int num : nums){
            Set<Integer> dp_new = new HashSet<>();
            for (int d : dp){
                dp_new.add(d);
                dp_new.add(d + num);
            }
            dp = dp_new;
        }
        if (dp.contains(summ/2)){
            return true;
        }
        else{
            return false;
        }
    }
}